src/python/grpcio_tests/tests_aio/unit/server_test.py

```diff
@@ -15,46 +15,211 @@
 from src.proto.grpc.testing import messages_pb2
 from src.proto.grpc.testing import benchmark_service_pb2_grpc
 from tests_aio.unit._test_base import AioTestBase
+from tests.unit.framework.common import test_constants
 
-_TEST_METHOD_PATH = ''
+_SIMPLE_UNARY_UNARY = '/test/SimpleUnaryUnary'
+_BLOCK_FOREVER = '/test/BlockForever'
+_BLOCK_BRIEFLY = '/test/BlockBriefly'
 
 _REQUEST = b'\x00\x00\x00'
 _RESPONSE = b'\x01\x01\x01'
 
 
-async def unary_unary(unused_request, unused_context):
-    return _RESPONSE
+class _GenericHandler(grpc.GenericRpcHandler):
 
+    def __init__(self):
+        self._called = asyncio.get_event_loop().create_future()
 
-class GenericHandler(grpc.GenericRpcHandler):
+    @staticmethod
+    async def _unary_unary(unused_request, unused_context):
+        return _RESPONSE
 
-    def service(self, unused_handler_details):
-        return grpc.unary_unary_rpc_method_handler(unary_unary)
+    async def _block_forever(self, unused_request, unused_context):
+        await asyncio.get_event_loop().create_future()
+
+    async def _BLOCK_BRIEFLY(self, unused_request, unused_context):
+        await asyncio.sleep(test_constants.SHORT_TIMEOUT / 2)
+        return _RESPONSE
+
+    def service(self, handler_details):
+        self._called.set_result(None)
+        if handler_details.method == _SIMPLE_UNARY_UNARY:
+            return grpc.unary_unary_rpc_method_handler(self._unary_unary)
+        if handler_details.method == _BLOCK_FOREVER:
+            return grpc.unary_unary_rpc_method_handler(self._block_forever)
+        if handler_details.method == _BLOCK_BRIEFLY:
+            return grpc.unary_unary_rpc_method_handler(self._BLOCK_BRIEFLY)
+
+    async def wait_for_call(self):
+        await self._called
+
+
+async def _start_test_server():
+    server = aio.server()
+    port = server.add_insecure_port('[::]:0')
+    generic_handler = _GenericHandler()
+    server.add_generic_rpc_handlers((generic_handler,))
+    await server.start()
+    return 'localhost:%d' % port, server, generic_handler
 
 
 class TestServer(AioTestBase):
 
     def test_unary_unary(self):
 
         async def test_unary_unary_body():
-            server = aio.server()
-            port = server.add_insecure_port('[::]:0')
-            server.add_generic_rpc_handlers((GenericHandler(),))
-            await server.start()
+            result = await _start_test_server()
+            server_target = result[0]
 
-            async with aio.insecure_channel('localhost:%d' % port) as channel:
-                unary_call = channel.unary_unary(_TEST_METHOD_PATH)
+            async with aio.insecure_channel(server_target) as channel:
+                unary_call = channel.unary_unary(_SIMPLE_UNARY_UNARY)
                 response = await unary_call(_REQUEST)
                 self.assertEqual(response, _RESPONSE)
 
         self.loop.run_until_complete(test_unary_unary_body())
 
+    def test_shutdown(self):
+
+        async def test_shutdown_body():
```

Comment:
Should we merge both tests to only once `test_shutdown` and `test_shutdown_after_call` and explicitly test that once the shutdown has been done new RPC calls fail because there is no server for handling the request?

Reply:
Personally, I separate them because it might give us insight when viewing the test results. E.g. `test_shutdown_after_call` failed but `test_shutdown` success, it might indicate some left-over from RPC is impairing the normal execution of the shutdown path. (Since most logic is finished developing, so such error is unlikely to appear. I'm also okay to merge them if you feel strongly.)

Reply:
should we then add a new test called `test_call_after_shutdown`? where the call fails - which is good - because there was a shutdown? Otherwise, we are not really testing that the server is shutdown.

Reply:
It is added at line 210.

Reply:
you are right!